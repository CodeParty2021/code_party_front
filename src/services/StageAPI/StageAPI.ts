import { createSlice } from "@reduxjs/toolkit";
import axios from "axios";

//TODO:ホストURLを環境変数で設定したい
const API_HOST = "http://localhost:8000";

export type Stage = {
  id: number;
  name: string;
  stage_index: number;
  rule: string;
};

export type Stages = {
  stageList: { [id: number]: Stage };
};

const stageSlice = createSlice({
  name: "stages",
  initialState: { stageList: {} } as Stages,
  reducers: {
    update: (state, action: { type: string; payload: Stage[] }) => {
      action.payload.forEach((v) => {
        state.stageList[v.id] = v;
      });
    },
    clear: (state) => {
      state.stageList = {};
    },
  },
});

export const { update, clear } = stageSlice.actions;

export const stageUpdateAsync = () => {
  return async (dispatch: any) => {
    axios
      .get(API_HOST + "/stages/?format=json")
      .then((res) => {
        const stages: Stage[] = res.data;
        dispatch(update(stages));
      })
      .catch();
  };
};

export const oneStageUpdateAsync = (id: number) => {
  return async (dispatch: any) => {
    axios
      .get(API_HOST + "/stages/" + id + "?format=json")
      .then((res) => {
        const stage: Stage = res.data;
        dispatch(update([stage]));
      })
      .catch();
  };
};

export default stageSlice.reducer;
