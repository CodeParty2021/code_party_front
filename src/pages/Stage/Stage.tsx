import React, { useEffect } from "react";
import { Link } from "react-router-dom";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store";
import { oneStageUpdateAsync } from "../../services/StageAPI/StageAPI";
import StageDetail from "./components/StageDetail";

type Prop = {};

export const Stage: React.FC<Prop> = () => {
  const stages = useSelector((state: RootState) => state.stages);
  const dispatch = useDispatch();
  const { id } = useParams();

  var info: React.ReactElement;

  if (id !== undefined) {
    var parsed = parseInt(id);
    //ステージ情報を取得
    info = <StageDetail stage={stages.stageList[parsed]}></StageDetail>;

    // ステージ情報を更新
    useEffect(() => {
      dispatch(oneStageUpdateAsync(parsed));
    }, [dispatch]);
  } else {
    //除外
    info = <>ステージ情報が見つかりませんでした．</>;
  }

  return (
    <div>
      <h1>Stage詳細</h1>
      <ul>
        <li>{info}</li>
      </ul>
      <Link to="/stage">ステージ一覧に戻る</Link>
    </div>
  );
};
