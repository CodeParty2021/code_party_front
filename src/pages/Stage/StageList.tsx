import React, { useEffect } from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store";
import { stageUpdateAsync } from "../../services/StageAPI/StageAPI";
import StageDetail from "./components/StageDetail";

type Prop = {};

export const StageList: React.FC<Prop> = () => {
  var list: React.ReactElement[] = [];

  const stages = useSelector((state: RootState) => state.stages);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(stageUpdateAsync());
  }, [dispatch]);

  Object.entries(stages.stageList).forEach(([key, value]) => {
    list.push(
      <li key={key}>
        <StageDetail stage={value}></StageDetail>
        <Link to={"/stage/" + value.id}>詳細を表示</Link>
      </li>
    );
  });

  return (
    <div>
      <h1>Stages</h1>
      <ul>{list}</ul>
      <Link to="/">トップに戻る</Link>
    </div>
  );
};
