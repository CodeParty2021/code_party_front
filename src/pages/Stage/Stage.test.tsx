import React from "react";
import { shallow } from "enzyme";
import { useSelector, useDispatch } from "react-redux";
import { Stage } from "./Stage";
import { Stages } from "../../services/StageAPI/StageAPI";
import { Params, useParams } from "react-router-dom";

jest.mock("react-redux");
jest.mock("react-router-dom");
jest.mock("../../services/StageAPI/StageAPI");

const useSelectorMock = useSelector as jest.Mock<Stages>;
const useDispatchMock = useDispatch as jest.Mock;
const useParamsMock = useParams as jest.Mock<Readonly<Params<string>>>;

describe("<Stage />", () => {
  beforeEach(() => {
    useSelectorMock.mockReturnValue({
      stageList: {
        0: {
          id: 0,
          name: "テストステージ",
          stage_index: 3,
          rule: "テストテキスト",
        },
      },
    });
    useDispatchMock.mockReturnValue(jest.fn());
    useParamsMock.mockReturnValue({ id: "0" });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("dispatches add stage action", () => {
    const wrapper = shallow(<Stage />);

    //expect(wrapper.find("li").text()).toEqual("<StageDetail />");
    expect(wrapper.getElements()).toMatchSnapshot();
  });
});
