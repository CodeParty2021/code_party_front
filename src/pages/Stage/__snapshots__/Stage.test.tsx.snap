// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stage /> dispatches add stage action 1`] = `
Array [
  <div>
    <h1>
      Stage詳細
    </h1>
    <ul>
      <li>
        <StageDetail
          stage={
            Object {
              "id": 0,
              "name": "テストステージ",
              "rule": "テストテキスト",
              "stage_index": 3,
            }
          }
        />
      </li>
    </ul>
    <ForwardRef(Link)
      to="/stage"
    >
      ステージ一覧に戻る
    </ForwardRef(Link)>
  </div>,
]
`;
